File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        rocks = Arrays.copyOf(rocks, rocks.length + 1);
        rocks[rocks.length - 1] = distance;
        Arrays.sort(rocks);
        
        int start = 1;
        int end = distance + 1;
        
        while((end - start) > 1) {
            int mid = (start + end) / 2;
            
            if(isValid(mid, rocks, n)) start = mid;
            else end = mid;
        }
        
        
        return start;
    }
    
    static boolean isValid(int d, int[] rocks, int n) {
        int removed = 0;
        int last = 0;
        
        for(int rock: rocks) {
            if(rock - last < d) {
                removed += 1;
                continue;
            }
            
            last = rock;
        }
        
        return removed <= n;
    }
}

File: 프로그래머스/4/43236. 징검다리/징검다리.py
"""
distance := 거리
rocks := 바위 위치
n := 제거할 바위의 수

answer := 최소 거리의 최댓값

모든 지점 사이의 거리가 d이상인 거리의 최댓값.
d를 중심으로 이진탐색..
"""
def isValid(d, rocks, n):
    removed = 0
    last = 0
    
    for rock in rocks:
        if rock - last < d:
            removed += 1
            continue
            
        last = rock
    return removed <= n

def solution(distance, rocks, n):
    start, end = 1, distance + 1
    rocks.append(distance)
    rocks.sort()
    
    while (end - start) > 1:
        mid = (start + end) // 2
        
        if isValid(mid, rocks, n):
            start = mid
        else:
            end = mid
        
    return start